openmc/lattice.py

```diff
@@ -0,0 +1,1901 @@
+                        universe.create_xml_subelement(xml_element)
+
+                # Get a string representation of the universe IDs.
+                slices.append(self._repr_axial_slice(self._universes[z]))
+
+            # Collapse the list of axial slices into a single string.
+            universe_ids = '\n'.join(slices)
+
+        # 2D Lattices
+        else:
+            # Initialize the center universe.
+            universe = self._universes[-1][0]
+            universe.create_xml_subelement(xml_element)
+
+            # Initialize the remaining universes.
+            for r in range(self._num_rings - 1):
+                for theta in range(6*(self._num_rings - 1 - r)):
+                    universe = self._universes[r][theta]
+                    universe.create_xml_subelement(xml_element)
+
+            # Get a string representation of the universe IDs.
+            universe_ids = self._repr_axial_slice(self._universes)
+
+        universes = ET.SubElement(lattice_subelement, "universes")
+        universes.text = '\n' + universe_ids
+
+        # Append the XML subelement for this Lattice to the XML element
+        xml_element.append(lattice_subelement)
+
+    @classmethod
+    def from_xml_element(cls, elem, get_universe):
+        """Generate hexagonal lattice from XML element
+
+        Parameters
+        ----------
+        elem : xml.etree.ElementTree.Element
+            `<hex_lattice>` element
+        get_universe : function
+            Function returning universe (defined in
+            :meth:`openmc.Geometry.from_xml`)
+
+        Returns
+        -------
+        HexLattice
+            Hexagonal lattice
+
+        """
+        lat_id = int(get_text(elem, 'id'))
+        name = get_text(elem, 'name')
+        lat = cls(lat_id, name)
+        lat.center = [float(i) for i in get_text(elem, 'center').split()]
+        lat.pitch = [float(i) for i in get_text(elem, 'pitch').split()]
+        lat.orientation = get_text(elem, 'orientation', 'y')
+        outer = get_text(elem, 'outer')
+        if outer is not None:
+            lat.outer = get_universe(int(outer))
+
+        # Get nested lists of universes
+        lat._num_rings = n_rings = int(get_text(elem, 'n_rings'))
+        lat._num_axial = n_axial = int(get_text(elem, 'n_axial', 1))
+
+        # Create empty nested lists for one axial level
+        univs = [[None for _ in range(max(6*(n_rings - 1 - r), 1))]
+                 for r in range(n_rings)]
+        if n_axial > 1:
+            univs = [deepcopy(univs) for i in range(n_axial)]
+
+        # Get flat array of universes
+        uarray = np.array([get_universe(int(i)) for i in
+                           get_text(elem, 'universes').split()])
+
+        # Fill nested lists
+        j = 0
+        for z in range(n_axial):
+            # Get list for a single axial level
+            axial_level = univs[z] if n_axial > 1 else univs
+
+            if lat.orientation == 'y':
```

Comment:
Consider splitting the X and Y functionality out into their own subclasses similarly to the `openmc.surface` module. This will reduce the complexity of many of these methods.